### PR TITLE
[EKA-35] Setup rabbitmq. Broadcasting and receiving consent artefact created message

### PR DESCRIPTION
**Pull request by chethandeshpande**

### CODE REVIEW

File: src/main/java/in/projecteka/consentmanager/ConsentManagerConfiguration.java [hunk 1]
```diff
@@ -128,6 +139,45 @@ public ConsentManager consentRequestService(WebClient.Builder builder,
...
 
+    @Bean
+    public DestinationsConfig destinationsConfig(AmqpAdmin amqpAdmin) {
+        HashMap<String, DestinationsConfig.DestinationInfo> queues = new HashMap<>();
+        queues.put(HIU_NOTIFICATION_QUEUE, new DestinationsConfig.DestinationInfo("exchange", HIU_NOTIFICATION_QUEUE));
+
+        DestinationsConfig destinationsConfig = new DestinationsConfig(queues, null);
+        destinationsConfig.getQueues()
+                .forEach((key, destination) -> {
+                    Exchange ex = ExchangeBuilder.directExchange(
```
Comment (angshu) [inline src/main/java/in/projecteka/consentmanager/ConsentManagerConfiguration.java]: These exchanges should not be created like this. 
- Can we first check if the exchange exists? I think there is a flag called auto-declare or something similar
- ideally these exchanges would be created externally.  If not handling that now, please create a tech debt
- keep in mind that in a cluster, CM grant service can be multiple, and not a single one

File: src/main/java/in/projecteka/consentmanager/consent/notification/ConsentArtefactBroadcastListener.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    @Autowired
+    private MessageListenerContainerFactory messageListenerContainerFactory;
+    @Autowired
+    private DestinationsConfig destinationsConfig;
+    @Autowired
+    private Jackson2JsonMessageConverter converter;
+
+    @PostConstruct
+    public void subscribe() {
+        DestinationsConfig.DestinationInfo destinationInfo = destinationsConfig
```
Comment (angshu) [inline src/main/java/in/projecteka/consentmanager/consent/notification/ConsentArtefactBroadcastListener.java]: What happens if the queue is not created by then? Ideally we should retry.

File: build.gradle [hunk 3]
```diff
@@ -41,6 +41,7 @@ task down(type: Exec) {
 
 dependencies {
 	implementation 'org.springframework.boot:spring-boot-starter-webflux'
+	compile group: 'org.springframework.boot', name: 'spring-boot-starter-amqp', version: '2.2.4.RELEASE'
```
Comment (ganesanarun) [inline build.gradle]: compile group: 'org.springframework.boot:spring-boot-starter-amqp'

File: src/main/java/in/projecteka/consentmanager/consent/ConsentManager.java [hunk 4]
```diff
@@ -34,6 +40,10 @@
     private final ConsentRequestRepository consentRequestRepository;
     private final ConsentArtefactRepository consentArtefactRepository;
     private KeyPair keyPair;
+    @Autowired
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/consent/ConsentManager.java]: Could you please remove @Autowired annotation, remove constructor and add @AllArgsConstructor?

File: src/main/java/in/projecteka/consentmanager/consent/ConsentManager.java [hunk 5]
```diff
@@ -98,19 +108,62 @@ private String getHIUId(ConsentDetail consentDetail) {
 
     public Mono<ConsentApprovalResponse> approveConsent(String authorization,
                                                         String requestId,
-                                                        List<GrantedConsent> grantedConsents) {
+                                                        List<GrantedConsent> grantedConsents,
+                                                        String callBackUrl) {
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/consent/ConsentManager.java]: From where we are getting the callBackUrl?

File: src/main/java/in/projecteka/consentmanager/consent/ConsentManager.java [hunk 6]
```diff
@@ -98,19 +108,62 @@ private String getHIUId(ConsentDetail consentDetail) {
...
         String patientId = TokenUtils.readUserId(authorization);
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/consent/ConsentManager.java]: PatientID should have been sent from controller, follow the same thing for other methods.

File: src/main/java/in/projecteka/consentmanager/consent/ConsentManager.java [hunk 7]
```diff
@@ -98,19 +108,62 @@ private String getHIUId(ConsentDetail consentDetail) {
...
-                                            .thenReturn(Consent.builder().id(consentArtefact.getId()).build());
-                                }).collectList())
-                .map(consents -> ConsentApprovalResponse.builder().consents(consents).build());
+                        generateConsentArtefacts(requestId, grantedConsents, patientId, consentRequest))
+                .flatMap(consents -> {
+                    ConsentApprovalResponse consentApprovalResponse = ConsentApprovalResponse.builder()
+                            .consents(consents)
+                            .build();
+                    return broadcastConsentArtefacts(callBackUrl, consentApprovalResponse.getConsents())
+                            .thenReturn(consentApprovalResponse);
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/consent/ConsentManager.java]: we could abstract out this part by having a class or something, PostConsentApproval?

File: src/main/java/in/projecteka/consentmanager/consent/ConsentManager.java [hunk 8]
```diff
@@ -98,19 +108,62 @@ private String getHIUId(ConsentDetail consentDetail) {
...
+            System.out.println("destination config is empty");
+            return Mono.empty();
+        }
+
+        return Mono.create(monoSink -> {
+            amqpTemplate.convertAndSend(
+                    destinationInfo.getExchange(),
+                    destinationInfo.getRoutingKey(),
+                    ConsentArtefactsNotification.builder().callBackUrl(callBackUrl).consents(consents).build());
+            System.out.println("Broadcasting consent artefact creation");
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/consent/ConsentManager.java]: :)

File: src/main/java/in/projecteka/consentmanager/consent/ConsentManager.java [hunk 9]
```diff
@@ -98,19 +108,62 @@ private String getHIUId(ConsentDetail consentDetail) {
...
+                patientId,
+                consentArtefactSignature);
+    }
+
+    @SneakyThrows
+    private Mono<Void> broadcastConsentArtefacts(String callBackUrl, List<Consent> consents) {
+        DestinationsConfig.DestinationInfo destinationInfo = destinationsConfig.getQueues().get(HIU_NOTIFICATION_QUEUE);
+
+        if (destinationInfo == null) {
+            System.out.println("destination config is empty");
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/consent/ConsentManager.java]:
```suggestion
```

:)

File: src/main/resources/static/api.yaml [hunk 10]
```diff
@@ -730,6 +730,9 @@ components:
             $ref: '#/components/schemas/ConsentApproveRequest'
           xml:
             wrapped: true
+        callBackUrl:
+          type: string
+          example: 'http://tmh-hiu.com/'
```
Comment (ganesanarun) [inline src/main/resources/static/api.yaml]:
```suggestion
          example: 'https://tmh-hiu.com/notify'
```

Comment (ganesanarun) [reply src/main/resources/static/api.yaml]: It should not be part of approve request no? ConsentRequest only should have it no?

File: src/main/java/in/projecteka/consentmanager/consent/notification/ConsentArtefactBroadcastListener.java [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+                .getQueues()
+                .get(HIU_NOTIFICATION_QUEUE);
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/consent/notification/ConsentArtefactBroadcastListener.java]: CONSENT_GRANTED_QUEUE?

File: src/test/java/in/projecteka/consentmanager/AuthorizationTest.java [hunk 12]
```diff
@@ -23,6 +24,9 @@
 @AutoConfigureWebTestClient
 public class AuthorizationTest {
 
+    @MockBean
+    private DestinationsConfig destinationsConfig;
+
```
Comment (ganesanarun) [inline src/test/java/in/projecteka/consentmanager/AuthorizationTest.java]: Why do we need this?

File: src/main/java/in/projecteka/consentmanager/MessageListenerContainerFactory.java [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+public class MessageListenerContainerFactory {
+
+    @Autowired
+    private ConnectionFactory connectionFactory;
+
+    @Autowired
+    private Jackson2JsonMessageConverter converter;
+
+    public MessageListenerContainer createMessageListenerContainer(String queueName) {
+        SimpleMessageListenerContainer container = new SimpleMessageListenerContainer(connectionFactory);
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/MessageListenerContainerFactory.java]: Can we not have @component and @autowired if possible?

File: src/main/java/in/projecteka/consentmanager/consent/model/request/ConsentApprovalRequest.java [hunk 14]
```diff
@@ -11,4 +11,7 @@
     @Valid
     @NotNull(message = "Consents are not specified")
     private List<GrantedConsent> consents;
+    @Valid
+    @NotNull(message = "Callback Url not specified")
+    private String callBackUrl;
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/consent/model/request/ConsentApprovalRequest.java]: Why it's part of ConsentApprovalRequest? shouldn't it be part of ConsentRequest from HIU?

File: src/main/java/in/projecteka/consentmanager/consent/ConsentManager.java [hunk 15]
```diff
@@ -102,18 +112,60 @@ private String getHIUId(RequestedDetail requestedDetail) {
...
-                                }).collectList())
-                .map(consents -> ConsentApprovalResponse.builder().consents(consents).build());
+                        generateConsentArtefacts(requestId, grantedConsents, patientId, consentRequest))
+                .flatMap(consents -> {
+                    ConsentApprovalResponse consentApprovalResponse = ConsentApprovalResponse.builder()
+                            .consents(consents)
+                            .build();
+                    return broadcastConsentArtefacts("callBackUrl", consentApprovalResponse.getConsents())
+                            .thenReturn(consentApprovalResponse);
+                });
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/consent/ConsentManager.java]: we could abstract out this part by having a class or something, PostConsentApproval, then consent manager.class doesn't need to know about amqb and all that.

File: src/main/java/in/projecteka/consentmanager/ConsentManagerConfiguration.java (hunk 1 above)
Comment (chethandeshpande) [reply src/main/java/in/projecteka/consentmanager/ConsentManagerConfiguration.java]: I have created a card to handle all this. 
https://trello.com/c/JmHRctGg

File: src/main/resources/static/api.yaml (hunk 10 above)
Comment (ganesanarun) [reply src/main/resources/static/api.yaml]: Can you remove the callback url from consent approve request? and added it to the consentrequest?

**Pull request merged**